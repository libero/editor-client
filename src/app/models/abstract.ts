import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';

import { makeSchemaFromConfig } from 'app/models/utils';
import * as abstractConfig from 'app/models/config/abstract.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter abstract')]
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter impact statement')]
  });
}
