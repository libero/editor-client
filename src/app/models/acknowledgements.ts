import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { Step } from 'prosemirror-transform';
import { gapCursor } from 'prosemirror-gapcursor';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from './utils';
import { PlaceholderPlugin } from './plugins/placeholder.plugin';
import { SelectionPlugin } from './plugins/selection.plugins';

export function createAcknowledgementsState(content?: Element, changeSteps?: [Step]): EditorState {
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter acknowledgements')]
  });
}
