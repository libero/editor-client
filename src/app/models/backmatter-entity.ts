import { v4 as uuidv4 } from 'uuid';

import { JSONObject } from '../types/utility.types';

export abstract class BackmatterEntity {
  constructor() {
    this._id = uuidv4();
  }

  get id(): string {
    return this._id;
  }

  protected _id: string;

  protected abstract fromXML(xmlNode: Element): void;
  protected abstract fromJSON(json: JSONObject): void;
  protected abstract createBlank(): void;

  protected createEntity(data?: Element | JSONObject): void {
    if (!data) {
      this.createBlank();
    } else if (data instanceof Element) {
      this.fromXML(data);
    } else {
      this.fromJSON(data);
    }
  }
}
