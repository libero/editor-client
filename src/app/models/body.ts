import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { set } from 'lodash';

import * as bodyConfig from './config/body.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from 'app/models/utils';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { createListKeymap } from 'app/utils/prosemirror/list.helpers';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createBodyState(content: Element, id: string): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  set(xmlContentDocument, 'manuscriptPath', `/manuscripts/${id}`);

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [
      buildInputRules(),
      gapCursor(),
      SelectionPlugin,
      dropCursor({ color: '#0078CF', class: 'drop-cursor' }),
      keymap(createListKeymap(schema)),
      keymap(baseKeymap),
      PlaceholderPlugin('Enter main text')
    ]
  });
}
