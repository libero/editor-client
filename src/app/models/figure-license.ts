import { getTextContentFromPath } from 'app/models/utils';

export interface FigureLicense {
  copyrightHolder: string;
  copyrightStatement: string;
  copyrightYear: string;
  licenseType: string;
}

export const FIGURE_LICENSE_CC0 = 'LICENSE_CCO';
export const FIGURE_LICENSE_CC_BY_4 = 'LICENSE_CC_BY_4';
export const FIGURE_LICENSE_OTHER = 'LICENSE_OTHER';

export const FIGURE_LICENSE_SELECT_OPTIONS = [
  { label: 'CCO', value: FIGURE_LICENSE_CC0 },
  { label: 'CC_BY_4', value: FIGURE_LICENSE_CC_BY_4 },
  { label: 'Other', value: FIGURE_LICENSE_CC_BY_4 }
];

const LICENSE_URLS_MAP = {
  'http://creativecommons.org/licenses/by/4.0/': FIGURE_LICENSE_CC_BY_4,
  'http://creativecommons.org/publicdomain/zero/1.0/': FIGURE_LICENSE_CC0
};

export function createFigureLicenseAttributes(el: Element): FigureLicense {
  return {
    copyrightHolder: getTextContentFromPath(el, 'copyright-holder'),
    copyrightStatement: getTextContentFromPath(el, 'copyright-statement'),
    copyrightYear: getTextContentFromPath(el, 'copyright-year'),
    licenseType: getLicenseType(el)
  };
}

export function createEmptyLicenseAttributes(): FigureLicense {
  return {
    copyrightHolder: '',
    copyrightStatement: '',
    copyrightYear: '',
    licenseType: undefined
  };
}

function getLicenseType(el: Element): string {
  const licenseUrl =
    getTextContentFromPath(el, 'license_ref') || el.querySelector('license').getAttribute('xlink:href') || '';

  return LICENSE_URLS_MAP[licenseUrl] || FIGURE_LICENSE_OTHER;
}
