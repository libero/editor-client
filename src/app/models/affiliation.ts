import { getTextContentFromPath } from 'app/models/utils';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Affiliation extends BackmatterEntity {
  label: string;
  institution: { name: string };
  address: { city: string };
  country: string;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  public getDisplayName(): string {
    return [this.institution.name, this.address.city, this.country].filter(Boolean).join(', ');
  }

  public clone(): Affiliation {
    const affitiation = new Affiliation();
    affitiation._id = this._id;
    affitiation.label = this.label;
    affitiation.country = this.country;
    affitiation.institution = { name: this.institution.name };
    affitiation.address = { city: this.address.city };
    return affitiation;
  }

  protected fromXML(xml: Element): void {
    this._id = xml.getAttribute('id');
    this.label = getTextContentFromPath(xml, 'label');
    this.institution = {
      name: [
        getTextContentFromPath(xml, 'institution[content-type="dept"]'),
        getTextContentFromPath(xml, 'institution:not([content-type])')
      ]
        .filter(Boolean)
        .join(', ')
    };

    this.address = {
      city: getTextContentFromPath(xml, 'addr-line named-content[content-type="city"]')
    };
    this.country = getTextContentFromPath(xml, 'country');
  }

  protected fromJSON(json: JSONObject): void {
    this._id = json.id as string;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }
}

export function createAffiliationsState(affiliationsXml: Element[]): Affiliation[] {
  return affiliationsXml.map((xml) => new Affiliation(xml));
}
