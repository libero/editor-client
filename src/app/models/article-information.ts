import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { buildInputRules } from './plugins/input-rules';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';

import { getTextContentFromPath, makeSchemaFromConfig } from './utils';
import * as licenseTextConfig from './config/license-text.config';
import { Person } from './person';
import moment from 'moment';
import { BackmatterEntity } from './backmatter-entity';
import { JSONObject } from '../types/utility.types';

export const LICENSE_CC_BY_4 = 'CC-BY-4';
export const LICENSE_CC0 = 'CC0';

const LICENSE_CC_BY_4_TEXT = `This article is distributed under the terms of the 
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/licenses/by/4.0/">
        Creative Commons Attribution License</ext-link>, which permits unrestricted use and redistribution provided that the original author 
    and source are credited.`;

const LICENSE_CC0_TEXT = `This is an open-access article, free of all copyright, and may be freely reproduced, 
    distributed, transmitted, modified, built upon, or otherwise used by anyone for any lawful purpose.
    The work is made available under the
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/publicdomain/zero/1.0/">Creative Commons CC0 public domain dedication</ext-link>.`;

export class ArticleInformation extends BackmatterEntity {
  public static getLicenseText(licenseType: string): EditorState {
    const paragraph = document.createElement('p');
    paragraph.innerHTML = {
      [LICENSE_CC_BY_4]: LICENSE_CC_BY_4_TEXT,
      [LICENSE_CC0]: LICENSE_CC0_TEXT
    }[licenseType];

    return ArticleInformation.createLicenseEditorState(paragraph);
  }

  public static createLicenseEditorState(content?: Element): EditorState {
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    const xmlContentDocument = document.implementation.createDocument('', '', null);

    if (content) {
      xmlContentDocument.appendChild(content);
    }

    return EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor()]
    });
  }

  articleType: string;
  dtd: string;
  articleDOI: string;
  publisherId: string;
  volume: string;
  elocationId: string;
  subjects: Array<string>;
  publicationDate: string;
  licenseType: string;
  copyrightStatement: string;
  licenseText: EditorState;

  constructor(data?: JSONObject | Element, authors?: Person[]) {
    super();
    this.createEntity(data);
    if (data instanceof Element && authors) {
      this.updateCopyrightStatementFromXml(data, authors);
    }
  }

  public clone(): ArticleInformation {
    const newArticleInformation = new ArticleInformation();
    newArticleInformation.articleType = this.articleType;
    newArticleInformation.dtd = this.dtd;
    newArticleInformation.articleDOI = this.articleDOI;
    newArticleInformation.elocationId = this.elocationId;
    newArticleInformation.volume = this.volume;
    newArticleInformation.publisherId = this.publisherId;
    newArticleInformation.subjects = [...this.subjects];
    newArticleInformation.copyrightStatement = this.copyrightStatement;
    newArticleInformation.licenseType = this.licenseType;
    newArticleInformation.publicationDate = this.publicationDate;
    newArticleInformation.licenseText = this.licenseText;

    return newArticleInformation;
  }

  public updateCopyrightStatement(authors: Person[]): void {
    if (this.licenseType === LICENSE_CC_BY_4) {
      const authorNamesSection =
        authors.length === 1
          ? authors[0].lastName
          : authors.length === 2
          ? `${authors[0].lastName} and ${authors[1].lastName}`
          : authors.length > 2
          ? `${authors[0].lastName} et al`
          : '';
      const date = moment(this.publicationDate);

      this.copyrightStatement = `Â© ${date.isValid() ? date.format('YYYY') : ''}, ${authorNamesSection}`;
    } else {
      this.copyrightStatement = '';
    }
  }

  protected fromXML(xmlNode: Element): void {
    this.articleType = getTextContentFromPath(xmlNode, 'article-meta subj-group[subj-group-type="heading"]');
    const articleEl = xmlNode.querySelector('article');
    this.dtd = articleEl ? articleEl.getAttribute('dtd-version') : '';
    this.articleDOI = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="doi"]');
    this.elocationId = getTextContentFromPath(xmlNode, 'article-meta elocation-id');
    this.volume = getTextContentFromPath(xmlNode, 'article-meta volume');
    this.publisherId = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="publisher-id"]');
    this.subjects = Array.from(xmlNode.querySelectorAll('subj-group[subj-group-type="major-subject"] subject')).map(
      (el: Element) => el.textContent
    );

    this.licenseType = this.getLicenseTypeFromXml(xmlNode);
    this.publicationDate = this.getPublicationDateFromXml(xmlNode);
    this.licenseText = ArticleInformation.createLicenseEditorState(
      xmlNode.querySelector('article-meta permissions license license-p')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this.articleType = json.articleType as string;
    this.dtd = json.dtd as string;
    this.articleDOI = json.articleDOI as string;
    this.elocationId = json.elocationId as string;
    this.volume = json.volume as string;
    this.publisherId = json.publisherId as string;
    this.subjects = (json.subjects as string[]) || [];

    this.licenseType = json.licenseType as string;
    this.publicationDate = json.publicationDate as string;
    const emptyEditorState = ArticleInformation.createLicenseEditorState();
    this.licenseText = EditorState.fromJSON(
      {
        schema: emptyEditorState.schema,
        plugins: emptyEditorState.plugins
      },
      json.licenseText as JSONObject
    );
  }

  protected createBlank(): void {
    this.articleType = '';
    this.dtd = '';
    this.articleDOI = '';
    this.elocationId = '';
    this.volume = '';
    this.publisherId = '';
    this.subjects = [];

    this.licenseType = '';
    this.publicationDate = '';
    this.licenseText = ArticleInformation.createLicenseEditorState();
  }

  private getLicenseTypeFromXml(doc: Element): string {
    const licenseEl = doc.querySelector('article-meta permissions license');
    if (!licenseEl) {
      return '';
    }
    const href = licenseEl.getAttribute('xlink:href');
    if (href === 'http://creativecommons.org/licenses/by/4.0/') {
      return LICENSE_CC_BY_4;
    }
    if (href === 'http://creativecommons.org/publicdomain/zero/1.0/') {
      return LICENSE_CC0;
    }
  }

  private getPublicationDateFromXml(xmlNode: Element): string {
    let publicationDate = '';
    const pubDateNode = xmlNode.querySelector('pub-date[date-type="pub"][publication-format="electronic"]');
    if (pubDateNode) {
      publicationDate = [
        pubDateNode.querySelector('year')?.textContent,
        pubDateNode.querySelector('month')?.textContent,
        pubDateNode.querySelector('day')?.textContent
      ].join('-');
    }

    return publicationDate;
  }

  private updateCopyrightStatementFromXml(xmlNode: Element, authors: Person[]): void {
    const copyrightStatementFromXml = getTextContentFromPath(xmlNode, 'article-meta permissions copyright-statement');
    if (copyrightStatementFromXml) {
      this.copyrightStatement = copyrightStatementFromXml;
    } else {
      this.updateCopyrightStatement(authors);
    }
  }
}
