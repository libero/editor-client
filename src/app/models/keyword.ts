import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';

import { makeSchemaFromConfig } from 'app/models/utils';
import * as keywordConfig from 'app/models/config/keywords.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export class Keyword extends BackmatterEntity {
  content: EditorState;

  constructor(data?: JSONObject | Element | EditorState) {
    super();
    if (data instanceof EditorState) {
      this.content = data;
    } else {
      this.createEntity(data);
    }
  }

  public clone(): Keyword {
    const kwd = new Keyword();
    kwd._id = this.id;
    kwd.content = this.content.apply(this.content.tr);
    return kwd;
  }

  protected fromXML(xmlNode: Element): void {
    const schema = makeSchemaFromConfig(keywordConfig.topNode, keywordConfig.nodes, keywordConfig.marks);
    this.content = EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlNode),
      schema,
      plugins: [buildInputRules(), SelectionPlugin, gapCursor(), dropCursor(), keymap(baseKeymap)]
    });
  }

  protected fromJSON(json: JSONObject): void {
    this._id = (json._id as string) || this._id;
    const blankState = this.createEmptyEditorState();
    this.content = EditorState.fromJSON(
      { schema: blankState.schema, plugins: blankState.schema.plugins },
      json.content as JSONObject
    );
  }

  protected createBlank(): void {
    this.content = this.createEmptyEditorState();
  }

  private createEmptyEditorState(): EditorState {
    const schema = makeSchemaFromConfig(keywordConfig.topNode, keywordConfig.nodes, keywordConfig.marks);
    return EditorState.create({
      doc: undefined,
      schema,
      plugins: [buildInputRules(), SelectionPlugin, gapCursor(), dropCursor(), keymap(baseKeymap)]
    });
  }
}

export interface KeywordGroup {
  title: string | undefined;
  keywords: Keyword[];
  newKeyword: Keyword;
}

export interface KeywordGroups {
  [keywordType: string]: KeywordGroup;
}

export function createKeywordGroupsState(keywordGroupsXml: Element[]): KeywordGroups {
  return keywordGroupsXml.reduce((acc, kwdGroup) => {
    const kwdGroupType = kwdGroup.getAttribute('kwd-group-type') || 'keywords-1';
    const groupTitle = kwdGroup.querySelector('title');
    const moreKeywords = Array.from(kwdGroup.querySelectorAll('kwd')).map(
      (keywordEl: Element, _) => new Keyword(keywordEl)
    );
    acc[kwdGroupType] = {
      title: groupTitle ? groupTitle.textContent : undefined,
      keywords: moreKeywords,
      newKeyword: new Keyword()
    };

    return acc;
  }, {});
}
