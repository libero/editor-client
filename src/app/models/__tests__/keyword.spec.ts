import { createKeywordGroupsState, createNewKeywordState } from 'app/models/keyword';

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

describe('Manuscript state factory', () => {
  it('creates keywords state', () => {
    const kwdContainer = document.createElement('div');
    kwdContainer.innerHTML = `<kwd-group kwd-group-type="author-keywords">
        <kwd>cerebellum</kwd>
        <kwd>climbing fiber</kwd>
        <kwd>reinforcement learning</kwd>
        <kwd><italic>vitae</italic></kwd>
        <kwd>Purkinje cells</kwd>
      </kwd-group>
      <kwd-group kwd-group-type="research-organism">
        <title>Research organism</title>
        <kwd>Mouse</kwd>
      </kwd-group>`;

    const editorState = createKeywordGroupsState(Array.from(kwdContainer.querySelectorAll('kwd-group')));
    expect(editorState).toMatchSnapshot();
  });

  it('creates empty keyword state', () => {
    expect(createNewKeywordState()).toMatchSnapshot();
  });
});
