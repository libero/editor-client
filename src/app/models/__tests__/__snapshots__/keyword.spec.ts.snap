// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manuscript state factory creates empty keyword state 1`] = `
Keyword {
  "_id": "unique_id",
  "content": Object {
    "doc": Object {
      "type": "keyword",
    },
    "selection": Object {
      "anchor": 0,
      "head": 0,
      "type": "text",
    },
  },
}
`;

exports[`Manuscript state factory creates keywords state 1`] = `
Object {
  "author-keywords": Object {
    "keywords": Array [
      Keyword {
        "_id": "unique_id",
        "content": Object {
          "doc": Object {
            "content": Array [
              Object {
                "text": "cerebellum",
                "type": "text",
              },
            ],
            "type": "keyword",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
      },
      Keyword {
        "_id": "unique_id",
        "content": Object {
          "doc": Object {
            "content": Array [
              Object {
                "text": "climbing fiber",
                "type": "text",
              },
            ],
            "type": "keyword",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
      },
      Keyword {
        "_id": "unique_id",
        "content": Object {
          "doc": Object {
            "content": Array [
              Object {
                "text": "reinforcement learning",
                "type": "text",
              },
            ],
            "type": "keyword",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
      },
      Keyword {
        "_id": "unique_id",
        "content": Object {
          "doc": Object {
            "content": Array [
              Object {
                "marks": Array [
                  Object {
                    "type": "italic",
                  },
                ],
                "text": "vitae",
                "type": "text",
              },
            ],
            "type": "keyword",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
      },
      Keyword {
        "_id": "unique_id",
        "content": Object {
          "doc": Object {
            "content": Array [
              Object {
                "text": "Purkinje cells",
                "type": "text",
              },
            ],
            "type": "keyword",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
      },
    ],
    "newKeyword": Keyword {
      "_id": "unique_id",
      "content": Object {
        "doc": Object {
          "type": "keyword",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
    },
    "title": undefined,
  },
  "research-organism": Object {
    "keywords": Array [
      Keyword {
        "_id": "unique_id",
        "content": Object {
          "doc": Object {
            "content": Array [
              Object {
                "text": "Mouse",
                "type": "text",
              },
            ],
            "type": "keyword",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
      },
    ],
    "newKeyword": Keyword {
      "_id": "unique_id",
      "content": Object {
        "doc": Object {
          "type": "keyword",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
    },
    "title": "Research organism",
  },
}
`;
